crates/storage/libmdbx-rs/src/txn_manager.rs

```diff
@@ -239,22 +196,38 @@ mod read_transactions {
                     // Iterate through active read transactions and abort those that's open for
                     // longer than `self.max_duration`.
                     for entry in self.active.iter() {
-                        let (ptr, start) = entry.pair();
+                        let (tx, start) = entry.value();
                         let duration = now - *start;
 
                         if duration > self.max_duration {
-                            let ptr = *ptr as *mut ffi::MDBX_txn;
-
-                            // Add the transaction to the list of aborted transactions, so further
-                            // usages report the correct error when the transaction is closed.
-                            self.add_aborted(ptr);
-
-                            // Abort the transaction
-                            let result = mdbx_result(unsafe { ffi::mdbx_txn_abort(ptr) });
-
-                            // Add the transaction to `aborted_active`. We can't remove it instantly
-                            // from the list of active transactions, because we iterate through it.
-                            aborted_active.push((ptr, duration, result.err()));
+                            let result = tx.txn_execute(|txn_ptr| {
+                                (
+                                    txn_ptr,
+                                    duration,
+                                    // Abort the transaction.
+                                    //
+                                    // We use `mdbx_txn_reset` instead of `mdbx_txn_abort` here to
+                                    // prevent MDBX from reusing the pointer of the aborted
+                                    // transaction for new read-only transactions. This is
+                                    // important because we store the pointer in the `active` list
+                                    // and assume that it is unique.
+                                    //
+                                    // See https://erthink.github.io/libmdbx/group__c__transactions.html#gae9f34737fe60b0ba538d5a09b6a25c8d for more info.
+                                    mdbx_result(unsafe { ffi::mdbx_txn_reset(txn_ptr) }),
```

Comment:
this will now create many dangling refs if user doesn't try to use again. an lru cache or similar map to `active` is needed to abort these (abort as in mdbx abort, not reset) eventually.

Reply:
I was thinking about it, but then it means that if the user eventually tries to use or abort it, we will report a EBADSIGN error and don't say anything about timeout. I think such bookkeeping for timeouted transactions is fine, especially if we'll add a metric for the number of timeouted transactions that weren't yet aborted by the user (so still has dangling pointer)

Reply:
yes, agreed, right direction here is to first add a metric of how many timed out txns mdbx can handle before crashing cause it runs out.

Reply:
runs out of ptrs

Reply:
I think it crashes when it runs out of memory, so if there's a leak of transactions. But if there's a leak, it's a bigger problem than just dangling references and timeouts shouldn't with a proper cleanup be the solution here.

Reply:
it could crash if there are no more pointers left to assign cause they are all locked by timeout and user doesn't try to use again

Reply:
just add a metric as mentioned and we're good to go

Reply:
> it could crash if there are no more pointers left to assign cause they are all locked by timeout and user doesn't try to use again

no more pointers left to assign = OOM, no? I may be wrong, but there's no hard limit on the number of transactions inside MDBX, and it's determined purely by available memory.

Reply:
sure, since it's on disk it's probably a lot of pointers though. still, in smaller envs, mdbx may be reliant on reassigning to perform well. maybe mdbx has some reserved pointers for ro txns. stuff we don't know about mdbx we discovered today, so what's the best thing we can do then? you said it, add metrics